src/base/LruMap.h

```diff
@@ -135,11 +153,29 @@ LruMap<EntryValue, EntryCost>::get(const char *key)
     return NULL;
 }
 
-template <class EntryValue, size_t EntryCost>
+template <class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
 bool
-LruMap<EntryValue, EntryCost>::add(const char *key, EntryValue *t)
+LruMap<EntryValue, MemoryUsedByEV>::add(const char *key, EntryValue *t)
 {
-    if (ttl == 0)
+    // map the retention concept into a TTL value
+
+    // do not store / retain any entries
+    if (retention == 0)
+        return false;
+
+    // negative retention means store 'forever'
+    if (retention < 0)
+        return add(key, t, 0xFFFFFFFF);
```

Comment:
Yes, you probably have mentioned it before. Perhaps I am still missing your point, but I still assert that a two-parameter add() method has full access to the retention value and should use it to form the right expiration date for the lower-level add_() or equivalent. Whatever logic the object creator has used to configure `retention` is irrelevant to the two-parameter add(). The two-parameter add() method just interprets the actual retention value as documented. I see no other sane way to implement the interface while adding an API to support caller-supplied TTL values.

The same overall approach applies to the addWithTtl() method or equivalent. There, the configured zero `retention` value overwrites supplied `ttl` (disabling all caching) and non-zero `retention` is ignored (favoring the explicit `ttl` value instead).

The fewer retention and ttl checks we have in the class, the better. With the approach I am outlining, most (possibly all?) such checks could be confined to the two top-level add() methods that convert either `this->retention` or the `ttl` parameter to the Entry expiration date used by the rest of the code (and to a "caching disabled" decision).

> Worst case I might have to go with dropping the magic number to 7-bit all-ones.

I cannot find a way to interpret the above "worst case" solution so that it works well:

1. An std::numeric_limits<int32_t>::max() is too big to be blindly added to `squid_curtime` (also a `time_t` value) safely.
1. A "7-bit all-ones number" (i.e., 127) is too small to represent "forever".

More importantly, I see no reason to make this complex or hacky. Just teach add() to compute the expiration date directly instead of computing 'ttl' that is then used to compute the expiration date! The third, internal add() method does not need to know anything about TTLs. It will only work with expiration dates supplied to it by the callers.

What is the largest expiration date stored as time_t? I think it is `std::numeric_limits<time_t>::max()`. You can even reduce code duplication (what if the Entry::expires type changes?) by using `decltype`!